### ISSUE TITLE
google_cloudfunctions2_function fails to deploy (error waiting for creating function, container healthcheck failed)

### ISSUE THREAD

**Issue by mgorsk1** [hashicorp/terraform-provider-google#13121]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

1.3.5

### Affected Resource(s)

* google_cloudfunctions2_function

### Terraform Configuration Files

```tf
# Source code ----------------------------------------------------------------------------------------------------------

# Creates a bucket to upload source code to.
resource "google_storage_bucket" "source" {
  name                        = "ramowka-357-schedule-code-bucket"
  location                    = "US"
  uniform_bucket_level_access = true
}

# Generates an archive of the source code compressed as a .zip file.
data "archive_file" "source" {
  type        = "zip"
  source_dir  = "../src"
  output_path = "/tmp/function.zip"
}

# Add source code zip to the Cloud Function's bucket.
resource "google_storage_bucket_object" "zip" {
  source       = data.archive_file.source.output_path
  content_type = "application/zip"

  # Append to the MD5 checksum of the files's content
  # to force the zip to be updated as soon as a change occurs
  name   = "src-${data.archive_file.source.output_md5}.zip"
  bucket = google_storage_bucket.source.name
}

# Function -------------------------------------------------------------------------------------------------------------

resource "google_cloudfunctions2_function" "default" {
  name     = "ramowka-357-schedule"
  location = "us-central1"

  build_config {
    runtime     = "python38"
    entry_point = var.function_name
    source {
      storage_source {
        bucket = google_storage_bucket.source.name
        object = google_storage_bucket_object.zip.name
      }
    }
  }

  service_config {
    max_instance_count = 1
    available_memory   = "128Mi"
    timeout_seconds    = 120

    environment_variables = {
      USER_EMAIL   = var.user_email
      SCHEDULE_URL = var.schedule_url
      DRY_RUN      = var.dry_run
    }
  }

  event_trigger {
    trigger_region = "us-central1"
    event_type     = "google.cloud.pubsub.topic.v1.messagePublished"
    pubsub_topic   = google_pubsub_topic.default.id
    retry_policy   = "RETRY_POLICY_RETRY"
  }

  depends_on = [
    google_pubsub_topic.default
  ]
}

# Trigger Job ----------------------------------------------------------------------------------------------------------

resource "google_pubsub_topic" "default" {
  name = "ramowka-357-schedule"
}

resource "google_cloud_scheduler_job" "default" {
  name     = "ramowka-357-schedule-sync-trigger"
  schedule = var.sync_schedule

  pubsub_target {
    topic_name = google_pubsub_topic.default.id
    data       = base64encode("trigger")
  }
}
```

### Debug Output

(https://gist.github.com/mgorsk1/79ab04e91e8ffe0d09e48ea563e67805)

### Expected Behavior

Function is deployed on GCP and triggered via pubsub message.

### Actual Behavior

Function appears in Google Cloud Console UI but with state Failed.

### Steps to Reproduce

1. `terraform apply`

### References

* #0000

**Comment by mgorsk1** [comment]
I think I found an issue, it's with `environment_variables` section, I changed this notation:

```
      USER_EMAIL   = var.user_email
      SCHEDULE_URL = var.schedule_url
      DRY_RUN      = var.dry_run
```

to this:

```
      APP_USER_EMAIL   = "${var.user_email}"
      APP_SCHEDULE_URL = "${var.schedule_url}"
      APP_DRY_RUN      = "${var.dry_run}"
```

and it worked. I am not sure what was the issue - name of the env variable or the notation I used (my suspicion is it's the former one). 

Assuming it's expected behavior I can close this issue. If not, I can reopen dedicated one @edwardmedia

**Comment by edwardmedia** [comment]
@mgorsk1 yeah I am glad you found how to fix it. I guess this is an issue more toward the Terraform core/sdk. Closing this now. Feel free to reopen if you want to continue the conversation

**Comment by mgorsk1** [comment]
Thanks @edwardmedia. I am not 100% sure on how to establish whether this is core/provider issue but I am now 100% sure the problem was with `DRY_RUN` key present inside `environment_variables` section. I deployed my stack multiple times now, including it and removing and every time it's there it's causing this problem (once removed/renamed the issue is gone). So I am for reopening it and I extended `Important Factoids` section with my discovery.
